src/main/java/com/oracle/nosql/spring/data/core/NosqlTemplateBase.java

```diff
@@ -387,4 +423,138 @@ private PreparedStatement getPreparedStatement(
     }
+
+    Map<String, FieldValue.Type> getShardKeys(Field idField) {
+        Map<String, FieldValue.Type> result = new LinkedHashMap<>();
+
+        if (NosqlEntityInformation.isCompositeKeyType(idField.getType())) {
+            Map<Integer, SortedSet<String>> shardKeys = new TreeMap<>();
+            NosqlPersistentEntity<?> compositeKeyEntity =
+                    mappingNosqlConverter.getMappingContext().
+                            getRequiredPersistentEntity(idField.getType());
+
+            for (NosqlPersistentProperty idProperty : compositeKeyEntity) {
+                if (idProperty.isWritable()) {
+                    if (idProperty.isAnnotationPresent(NosqlKey.class)) {
+                        NosqlKey noSqlKey =
+                                idProperty.findAnnotation(NosqlKey.class);
+                        int order = noSqlKey.order();
+                        if (noSqlKey.shardKey()) {
+                            SortedSet<String> ss = shardKeys.getOrDefault(order,
+                                    new TreeSet<>(String.CASE_INSENSITIVE_ORDER));
+                            ss.add(idProperty.getName());
+                            shardKeys.put(order, ss);
+                        }
+                    } else {
+                        SortedSet<String> ss =
+                                shardKeys.getOrDefault(NOTSET_PRIMARY_KEY_ORDER,
+                                        new TreeSet<>(String.CASE_INSENSITIVE_ORDER));
+                        ss.add(idProperty.getName());
+                        shardKeys.put(NOTSET_PRIMARY_KEY_ORDER, ss);
+                    }
+                }
+            }
+            List<String> sortedShardKeys = new ArrayList<>();
+
+            shardKeys.forEach((order, keys) -> {
+                //order should be specified on all fields if at all is used
+                if (order != NOTSET_PRIMARY_KEY_ORDER &&
+                        shardKeys.get(NOTSET_PRIMARY_KEY_ORDER) != null) {
+                    throw new IllegalArgumentException("If order is " +
+                            "specified, it must be specified on all shard key" +
+                            " fields of the composite key class " +
+                            compositeKeyEntity.getName());
+                }
+
+                //order value must be unique
+                if (order != NOTSET_PRIMARY_KEY_ORDER && keys.size() > 1) {
+                    throw new IllegalArgumentException("Order of " +
+                            "shard keys must be unique in composite key" +
+                            "class " + compositeKeyEntity.getName());
+                }
+                sortedShardKeys.addAll(keys);
+            });
+
+            if (sortedShardKeys.isEmpty()) {
+                throw new IllegalArgumentException("At least one of the " +
+                        "@NoSqlKey must be shard key in class " +
+                        compositeKeyEntity.getName());
+            }
+
```

Comment:
Need to check that comp field names don't collide with each other when case is ignored and they are not named kv_json_. This is valid for both shard and non-shard keys in the same time because they will end up as columns of the table.